### PR TITLE
Implemented exceptions aggregation for retriable executors.

### LINKED ISSUE #100: Provide additional useful information in thrown exception while retrying http request
https://github.com/reportportal/commons-net/blob/8ead01bd0ba1d123dabd6787892a168395a8e80b/src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs#L100

Would be nice to see in the final exception all attempts, timings, any other useful work made during retrying http requests.

### PR DESCRIPTION
**Pull request by Aleh-Yanushkevich**
@nvborisenko could you please review my ideas of catching intermediate exceptions? If you have any proposal, please write down here. 

Fixes #100

### CODE REVIEW

**Comment by Aleh-Yanushkevich** [comment]
I would like to clarify:

1. Should me add any message during throwing AggregateException?
2. Is it okay, that I've updated `IStatisticCounter`?
3. Maybe `RetryHttpRequestExeption` is not good name for intermediate exception.

**Comment by nvborisenko** [comment]
Let's agree here what final exception we would see.

```
System.AggregateException: What message we wanna to see here? (Attempt #1: Something happened. Next attempt in 5 seconds.) (Attempt #2 Something happened.)
 ---> System.Exception: Attempt #1: Something happened. Next attempt in 5 seconds.
   at Program.<Main>$(String[] args) in D:\Temp\ConsoleApp11\ConsoleApp11\Program.cs:line 8
   --- End of inner exception stack trace ---
   at Program.<Main>$(String[] args) in D:\Temp\ConsoleApp11\ConsoleApp11\Program.cs:line 26
 ---> (Inner Exception #1) System.Exception: Attempt #2 Something happened.
   at Program.<Main>$(String[] args) in D:\Temp\ConsoleApp11\ConsoleApp11\Program.cs:line 17<---
```

**Comment by nvborisenko** [comment]
I like how default `AggregateException` formats final string, but all inner messages are included into final message
```
System.AggregateException: What message we wanna to see here? (Attempt #1: Something happened. Next attempt in 5 seconds.) (Attempt #2 Something happened.)
```

Given that we have long message for failed http request, thus I don't like this feature. As I understood, we should introduce custom type to override this behaviour, and what is good type name...

File: src/ReportPortal.Shared/Reporter/Statistics/IStatisticsCounter.cs [hunk 1]
```diff
@@ -22,6 +22,11 @@ public interface IStatisticsCounter
         /// </summary>
         TimeSpan Avg { get; }
 
+        /// <summary>
+        /// Duration of last measured request.
```
Comment (nvborisenko) [inline src/ReportPortal.Shared/Reporter/Statistics/IStatisticsCounter.cs:26]: No needed here, this is just accumulated final stats.

Comment (Aleh-Yanushkevich) [reply src/ReportPortal.Shared/Reporter/Statistics/IStatisticsCounter.cs:26]: In this situation we need to duplicate logic of getting information about duration of request (if it's really needed for us). Right now it's implemented in base executor class and used by `IStatisticCounter`.

**Comment by Aleh-Yanushkevich** [comment]
> I like how default `AggregateException` formats final string, but all inner messages are included into final message
> 
> ```
> System.AggregateException: What message we wanna to see here? (Attempt #1: Something happened. Next attempt in 5 seconds.) (Attempt #2 Something happened.)
> ```
> 
> Given that we have long message for failed http request, thus I don't like this feature. As I understood, we should introduce custom type to override this behaviour, and what is good type name...

agree with it. Too much infomration will be. Maybe pass to final message details about last 3-5 exception + total count of attempts? I think that for information it will be enough.

File: src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs [hunk 2]
```diff
@@ -84,11 +87,12 @@ public override async Task<T> ExecuteAsync<T>(Func<Task<T>> func, Action<Excepti
...
                     Array.IndexOf(HttpStatusCodes, (exp as Client.ServiceException)?.HttpStatusCode) > -1)
                 {
+                    var delay = (int)Math.Pow(BaseIndex, i + MaxRetryAttemps);
+
                     if (i < MaxRetryAttemps - 1)
                     {
-                        var delay = (int)Math.Pow(BaseIndex, i + MaxRetryAttemps);
-
                         TraceLogger.Error($"Error while invoking '{func.Method.Name}' method. Current attempt: {i}. Waiting {delay} seconds and retrying it.\n{exp}");
+                        exceptions.Add(new HttpRequestException($"Something unexpected happened. Next attempt in {delay} second(s).", exp));
```
Comment (nvborisenko) [inline src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs:95]: We should be more specific here.

File: src/ReportPortal.Shared/Internal/Delegating/Exceptions/RetryExecutionException.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    public class RetryExecutionException : AggregateException
+    {
+        private readonly string _message;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="RetryExecutionException"/>
+        /// </summary>
+        /// <param name="innerExceptions">Inner exceptions.</param>
+        public RetryExecutionException(IEnumerable<Exception> innerExceptions)
+            : this("Request has not finished successfully", innerExceptions)
```
Comment (nvborisenko) [inline src/ReportPortal.Shared/Internal/Delegating/Exceptions/RetryExecutionException.cs:20]: `Request has not finished successfully` - let's be more generic here, it not only about http requests.

File: src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs [hunk 4]
```diff
@@ -97,7 +101,9 @@ public override async Task<T> ExecuteAsync<T>(Func<Task<T>> func, Action<Excepti
                     else
                     {
                         TraceLogger.Error($"Error while invoking '{func.Method.Name}' method. Current attempt: {i}.\n{exp}");
-                        throw;
+                        exceptions.Add(new HttpRequestException("Something unexpected happened. Limit of retries has been reached.", exp));
```
Comment (nvborisenko) [inline src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs:104]: `Something unexpected happened` - be more specific here as well.

File: src/ReportPortal.Shared/Internal/Delegating/Exceptions/RetryExecutionException.cs (hunk 3 above)
Comment (Aleh-Yanushkevich) [reply src/ReportPortal.Shared/Internal/Delegating/Exceptions/RetryExecutionException.cs]: fixed.

File: src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs (hunk 2 above)
Comment (Aleh-Yanushkevich) [reply src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs]: fixed.

File: src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs (hunk 4 above)
Comment (Aleh-Yanushkevich) [reply src/ReportPortal.Shared/Internal/Delegating/ExponentialRetryRequestExecuter.cs]: fixed.

**Review by nvborisenko — APPROVED** [review]
Thanks for that.

**Pull request merged**